d3a_interface/matching_algorithms/pay_as_clear_matching_algorithm.py

```diff
@@ -119,48 +122,48 @@ def _clearing_point_from_supply_demand_curve(bids_rate_energy: Dict, offers_rate
             if len(clearing) > 0:
                 return clearing[-1].rate, clearing[-1].energy
 
-    def get_clearing_point(self, bids: List[Dict], offers: List[Dict], current_time):
+    def get_clearing_point(self, bids: List[Dict], offers: List[Dict], current_time: DateTime,
+                           market_id: str) -> Union[Clearing, None]:
+        """Sorts Bids and Offers to find the equilibrium point"""
         self.sorted_bids = sort_list_of_dicts_by_attribute(bids, "energy_rate", True)
         self.sorted_offers = sort_list_of_dicts_by_attribute(offers, "energy_rate")
-        clearing = None
+        clearing, cumulative_bids, cumulative_offers = None, None, None
 
         if len(self.sorted_bids) == 0 or len(self.sorted_offers) == 0:
             return
 
         if ConstSettings.IAASettings.PAY_AS_CLEAR_AGGREGATION_ALGORITHM == 1:
             cumulative_bids = self._accumulated_energy_per_rate(self.sorted_bids)
             cumulative_offers = self._accumulated_energy_per_rate(self.sorted_offers)
-            self.state.cumulative_bids = cumulative_bids
-            self.state.cumulative_offers = cumulative_offers
             ascending_rate_bids = OrderedDict(reversed(list(cumulative_bids.items())))
             clearing = self._clearing_point_from_supply_demand_curve(
                 ascending_rate_bids, cumulative_offers)
         elif ConstSettings.IAASettings.PAY_AS_CLEAR_AGGREGATION_ALGORITHM == 2:
             cumulative_bids = self._discrete_point_curve(self.sorted_bids, math.floor)
             cumulative_offers = self._discrete_point_curve(self.sorted_offers, math.ceil)
-            max_rate = self._populate_market_cumulative_offer_and_bid(cumulative_bids,
-                                                                      cumulative_offers)
-            clearing = self._get_clearing_point(max_rate)
+            max_rate, cumulative_bids, cumulative_offers = \
+                self._populate_market_cumulative_offer_and_bid(cumulative_bids, cumulative_offers)
+            clearing = self._get_clearing_point(max_rate, cumulative_bids, cumulative_offers)
         if clearing is not None:
-            self.state.clearing[current_time] = clearing[0]
+            self.state.cumulative_bids[market_id] = {current_time: cumulative_bids}
+            self.state.cumulative_offers[market_id] = {current_time: cumulative_offers}
+            self.state.clearing[market_id] = {current_time: clearing}
```

Comment:
The `market_id` and these lines do not seem to belong to this method, as they don't have any effect on returning the `clearing_point` that you're returning. Also, they break the [Command-Query separation principle](https://en.wikipedia.org/wiki/Command%E2%80%93query_separation).
I would recommend moving them into a new method and call it from `get_matches_recommendations()`.

Reply:
Also, why do we need to save the `current_time` in there. Isn't the `market_id` a sufficient unique identifier for the market?
Sorry that I only ask this now.

Reply:
> Also, why do we need to save the current_time in there. Isn't the market_id a sufficient unique identifier for the market?
Sorry that I only ask this now.

There can be multiple clearings in a single market. Hope it answers your question.

Reply:
> The market_id and these lines do not seem to belong to this method, as they don't have any effect on returning the clearing_point that you're returning (and they also break the Command-Query separation principle).
I would recommend moving them into a new method and call it from get_matches_recommendations()

Moving it to a new method and calling it from `get_matches_recommendations` requires `get_clearing_point` to return additional values like `cumulative_offers` & `cumulative_bids` which violates the method name from which it expects only to return a clearing point. As target of this *bug ticket* was to solve a bug, not refactoring, I would rather create a ticket to think of better refactoring.